src/stdune/io.ml

```diff
@@ -17,6 +17,47 @@ let input_lines =
   in
   fun ic -> loop ic []
 
+let input_zero_from_buffer from buf =
+  match String.index_from_opt buf from '\x00' with
+  | None -> (None, from)
+  | Some eos -> (Some (String.sub buf ~pos:from ~len:(eos - from)), eos + 1)
+
+let input_zero_separated =
+  (* Take all the \0-terminated strings from [buf], return the scanned list and
+     the remainder *)
+  let rec scan_inputs_buf from buf acc =
+    match input_zero_from_buffer from buf with
+    | Some istr, from -> scan_inputs_buf from buf (istr :: acc)
+    | None, from ->
+      let total_len = String.length buf in
+      if total_len > from then
+        let rest = String.sub buf ~pos:from ~len:(total_len - from) in
+        (acc, Some rest)
+      else
+        (acc, None)
+  in
+  let ibuf_size = 65536 in
+  let ibuf = Bytes.create ibuf_size in
+  let rec input_loop ic rem acc =
+    let res = input ic ibuf 0 ibuf_size in
+    if res = 0 then
+      (* end of file, check if there is a remainder, and return the results *)
+      match rem with
+      | Some rem -> List.rev (rem :: acc)
+      | None -> List.rev acc
+    else
+      (* new input, append remainder and scan it *)
+      let actual_input = Bytes.sub_string ibuf ~pos:0 ~len:res in
+      let actual_input =
+        match rem with
+        | None -> actual_input
+        | Some rem -> rem ^ actual_input
```

Comment:
If we wanted to improve the worst-case scenario of repeatedly concatenating a string here, `rem` could be a `string list` instead (the function `scan_inputs_buf` would have to be adapted in turn).

But I understand that this is not relevant for the case at hand of dealing with filenames.

Reply:
Let's improve this indeed.

Reply:
After giving it a go IMHO the extra complexity of the code was not worth it; I've added a comment instead. We can improve it later if we want.

Reply:
Makes sense, thanks.